### ISSUE TITLE
Non-Autoregressive Neural Machine Translation

### ISSUE THREAD

**Issue by baoy-nlp** [baoy-nlp/Reader#1]
[https://github.com/kweonwooj/papers/issues/30#issue-269283193](url)

**Comment by baoy-nlp** [comment]
Abstract
Non-autoregressive NMT allows an order of magnitude lower latency during inference.
Through knowledge distillation, the use of input token fertilities as a latent variables, and policy gradient fine-tuning, they achieve 2.0 BLEU points lower than autoregressive Transformer network used as a teacher
screen shot 2017-10-28 at 1 08 22 pm

Details
Towards non-autoregressive decoding
naive method is to predict each output independently, but this does not yield good result due to multimodality problem, a conditional independence between various target sentences.
for example, 'thank you' in English can be translated into 'danke schon' or 'vielen dank', but 'danke dank' is not acceptable.
Non-Autoregressive Transformer (NAT)

Encoder stack

encoder stays unchanged from original Transformer
Decoder stack

Decoder inputs

copy source inputs using fertilities
Fertilities mean the number of times each input is copied into decoder inputs, it controls the "speed" at which decoder translates and determines the length of target sentence
Non-causal self-attention

since no masking is needed for later tokens, simply use self-attention with masking out each query position only from attending to itself
Positional attention
include additional positional attention module in decoder layer, which adds stronger positional signal and allows decoder to perform local reordering
Fertility

a latent variable that models the nondeterminism in the translation process
sample z from a prior distribution and then condition on z to non-autoregressively generate a translation
with max = 50
Conditional Probability of a Target Translation, Y is
screen shot 2017-10-28 at 1 20 12 pm

Translation Predictor and Decoding Process

searching for all combinations of fertility is intractable, so
choose argmax fertility for each input word
choose expected value for fertility
noisy parallel decoding, samples random fertility sequence and decode in parallel, then choose argmax
Training (read again)
screen shot 2017-10-28 at 1 23 02 pm

Sequence-level Knowledge Distillation
use teacher-generated target corpus for training, original corpus results in too noisy and nondeterministic
Fine-tuning stage with reverse KL divergence with teacher output distribution in a form of word-level knowledge distillation
KD is favorable towards highly peaked student output distributions than a standard cross-entropy error would be
screen shot 2017-10-28 at 1 25 37 pm
Joint training (read again)
sum of original distillation loss, expectation over fertility distribution, normalized with a baseline and other based on external fertility inference model via policy gradient and backprop
screen shot 2017-10-28 at 1 26 43 pm
Experiments

IWSLT16 En-De as development
WMT14 En-De and WMT16 En-Ro as final result verification
use shared BPE, for IWSLT use separate vocab and embedding
transfer student's encoder weights from encoder weights from its teacher
fertility prediction is supervised with fast_align by IBM Model 2 during training
Results
screen shot 2017-10-28 at 1 30 58 pm
screen shot 2017-10-28 at 1 34 23 pm

NAT 2~5 BLEU less than AT

Speed up for > x15 over beam search in teacher model
screen shot 2017-10-28 at 1 37 41 pm

Good experimental scheme

Fine-tune does not converge with RL or BP alone, must use all three fine-tuning terms to get +1.5 BLEU

Overall Structure
screen shot 2017-10-28 at 1 43 02 pm

src_len vs Latency
screen shot 2017-10-28 at 1 43 04 pm

learning curve for NAT (bleu on dev set)
screen shot 2017-10-28 at 1 43 09 pm
